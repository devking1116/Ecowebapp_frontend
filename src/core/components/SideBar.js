/* @flow */
import React from 'react';
import reactCSS from 'reactcss';
import { Menu, Icon } from 'semantic-ui-react';

type Props = {
  style?: Object,
  header: React.Element<*>,
  expandedWidth: number | string,
  isExpanded?: bool,
  hiddenWhenClosed?: bool,
  children?: any,
};

/**
 * A sidebar that fill up the height of the container.
 *
 * NOTE: The header and children will be passed the isExpanded prop given to <SideBar />.
 */
export default function SideBar(props: Props): React.Element<*> {
  const {
    style,
    header,
    expandedWidth,
    isExpanded,
    hiddenWhenClosed,
    children,
    ...otherProps,
  } = props;

  const styles = reactCSS({
    'default': {
      sideBar: {
        position: 'relative',
        height: '100%',
        width: undefined,
        ...style,
      },
    },
    'isExpanded': {
      sideBar: {
        width: expandedWidth,
      },
    },
    'isClosed': {
      sideBar: {
        display: hiddenWhenClosed? 'none': undefined,
      },
    },
  }, { isExpanded, isClosed: !isExpanded });

  return (
    <Menu style={styles.sideBar} icon={!isExpanded} vertical {...otherProps}>
      <Menu.Item>
        {React.cloneElement(header, { isExpanded: isExpanded? true: false })}
      </Menu.Item>
    {React.Children.map(children, (child) =>
      <Menu.Item>
        {React.cloneElement(child, { isExpanded: isExpanded? true: false })}
      </Menu.Item>
    )}
    </Menu>
  );
}
