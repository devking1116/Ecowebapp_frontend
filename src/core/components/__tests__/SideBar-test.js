import React from 'react';
import renderer from 'react-test-renderer';

import SideBar from '../SideBar';

test('when expanded, it displays its header / children in a single column at the given width.', () => {
  const component = renderer.create(
    <SideBar header={<header>Test Header</header>} expandedWidth={200} isExpanded>
      <i>Test Child 0</i>
      <i>Test Child 1</i>
      <i>Test Child 2</i>
    </SideBar>
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('when not expaned, it displays its header / children in a single column as thin as it can be.', () => {
  const component = renderer.create(
    <SideBar header={<header>Test Header</header>} expandedWidth={200} >
      <i>Test Child 0</i>
      <i>Test Child 1</i>
      <i>Test Child 2</i>
    </SideBar>
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('when not expaned and hiddenWhenClosed is set, it does not display (style is display: none).', () => {
  const component = renderer.create(
    <SideBar header={<header>Test Header</header>} hiddenWhenClosed>
      <i>Test Child 0</i>
      <i>Test Child 1</i>
      <i>Test Child 2</i>
    </SideBar>
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('when expanded, it gives its header / children the isExpanded prop.', () => {
  const component = renderer.create(
    <SideBar header={<header>Test Header</header>} isExpanded>
      <i>Test Child 0</i>
      <i>Test Child 1</i>
      <i>Test Child 2</i>
    </SideBar>
  );
  const tree = component.toJSON();
  const { children: [ { children: [ header ] }, ...childrenMenuItems ] } = tree;

  expect(header.props.isExpanded).toBe(true);

  childrenMenuItems.forEach(({ children: [ child ]}) => {
    expect(child.props.isExpanded).toBe(true);
  });
});

test('when not expanded, it does not gives its header / children the isExpanded prop.', () => {
  const component = renderer.create(
    <SideBar header={<header>Test Header</header>}>
      <i>Test Child 0</i>
      <i>Test Child 1</i>
      <i>Test Child 2</i>
    </SideBar>
  );
  const tree = component.toJSON();const { children: [ header, ...children ] } = tree;

  expect(header.props.isExpanded).toBeFalsy();

  children.forEach(({ children: [ child ]}) => {
    expect(child.props.isExpanded).toBeFalsy();
  });
});
