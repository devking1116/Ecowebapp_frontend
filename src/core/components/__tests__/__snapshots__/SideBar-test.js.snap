exports[`test when expanded, it displays its header / children in a single column at the given width. 1`] = `
<div
  className="ui vertical menu"
  style={
    Object {
      "height": "100%",
      "position": "relative",
      "width": 200,
    }
  }>
  <div
    className="item"
    onClick={[Function]}>
    <header
      isExpanded={true}>
      Test Header
    </header>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={true}>
      Test Child 0
    </i>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={true}>
      Test Child 1
    </i>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={true}>
      Test Child 2
    </i>
  </div>
</div>
`;

exports[`test when not expaned and hiddenWhenClosed is set, it does not display (style is display: none). 1`] = `
<div
  className="ui vertical icon menu"
  style={
    Object {
      "display": "none",
      "height": "100%",
      "position": "relative",
      "width": undefined,
    }
  }>
  <div
    className="item"
    onClick={[Function]}>
    <header
      isExpanded={false}>
      Test Header
    </header>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={false}>
      Test Child 0
    </i>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={false}>
      Test Child 1
    </i>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={false}>
      Test Child 2
    </i>
  </div>
</div>
`;

exports[`test when not expaned, it displays its header / children in a single column as thin as it can be. 1`] = `
<div
  className="ui vertical icon menu"
  style={
    Object {
      "display": undefined,
      "height": "100%",
      "position": "relative",
      "width": undefined,
    }
  }>
  <div
    className="item"
    onClick={[Function]}>
    <header
      isExpanded={false}>
      Test Header
    </header>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={false}>
      Test Child 0
    </i>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={false}>
      Test Child 1
    </i>
  </div>
  <div
    className="item"
    onClick={[Function]}>
    <i
      isExpanded={false}>
      Test Child 2
    </i>
  </div>
</div>
`;
